Documenting Cascading Order: Introduction and Examples (#1648)
withastro/docs: src/pages/en/guides/styling.md

Because this approach uses the `public/` directory, it skips the normal CSS processing, bundling and optimizations that are provided by Astro. If you need these transformations, use the [Import a Stylesheet](#import-a-local-stylesheet) method above.

## Cascading Order

> When evaluating multiple CSS rules that apply to the same element, the browser uses _specificity_ and _order of appearance_ to determine which style to show. If one rule is more _specific_ than another, it will take precedence:
> 
> ```astro title="MyComponent.astro"
> <style>
>   h1 { color: red }
>   div > h1 {
>     color: purple
>   }
> </style>
> <div>
>   <h1 class="red">
>     This header will be purple!
>   </h1>
> </div>
> ```
> 
> If two rules have the same specificity, the _order of appearance_ is evaluated, and the last rule will take precedence:
> ```astro title="MyComponent.astro"
> <style>
>   h1 { color: purple }
>   h1 { color: red }
> </style>
> <div>
>   <h1 class="red">
>     This header will be red!
>   </h1>
> </div>
> ```
> 
> Among Astro features, if the specificity of a rule is the same, the resulting styles will appear in this order:
> 
> - **`<link>` tags in the head** (lowest precedence)

Reviewer: This is not Astro-specific, so I think it might make sense to provide an MDN link to Cascading Order and make it a bit more clear that you're "teaching CSS" here, not Astro. And, I *might* even remove these initial examples altogether.

OR, I'd keep the examples, but then make the text more super-Astro specific:

``````suggestion
Astro components will sometimes have to evaluate multiple CSS "instructions" for the same element. For example, your component might import a CSS stylesheet, include its own `<style>` tag, *and* be rendered inside a layout.

When multiple CSS rules apply to the same element, browsers first use _specificity_ and then _order of appearance_ to determine which style to show.

If one rule is more _specific_ than another, no matter when the CSS rule is evaluated, it will take precedence:

```astro title="MyComponent.astro"
<style>
  h1 { color: red }
  div > h1 {
    color: purple
  }
</style>
<div>
  <h1>
    This header will be purple!
  </h1>
</div>
```

If two rules have the same specificity, then the _order of appearance_ is evaluated, and the last rule will take precedence:
```astro title="MyComponent.astro"
<style>
  h1 { color: purple }
  h1 { color: red }
</style>
<div>
  <h1>
    This header will be red!
  </h1>
</div>
```

Astro CSS rules are evaluated in this order of appearance:

- **`<link>` tags in the head** (lowest precedence)
``````
Also, I'm not sure about using `class="red"` because we never define what the red class is? Are we supposed to infer that the class has a styling rule to make *something* red? All text? just h1? The `<style>` tag in the component isn't styling the class, it's styling the element. 

It seems these examples work without the class applied to them, and if we're not actually showing where the class comes from, or that it's even being applied (we don't import a global stylesheet, nor do we show that this component is in a layout that does) then I don't think it's helping explain why the style tag is making the h1 red.

Author: Oh no, I accidentally left this in when I copied it from an experimentation - will remove! 🤦

- **imported styles**
- **scoped styles** (highest precedence)

### Scoped Styles 

Using [scoped styles](#scoped-styles) does not increase the _specificity_ of your CSS, but it will always come last in the _order of appearance_. They will therefore take precedence over other styles of the same specificity. For example, if you import a stylesheet with the same rule as a scoped style, the scoped style will apply:

> ```css title="make-it-purple.css"

Reviewer: I love these scoped style examples!! `make-it-so`  😂

h1 {
  color: purple;
}
```
```astro title="MyComponent.astro"
---
import "./make-it-purple.css"
---
<style>
  h1 { color: red }
</style>
<div>
  <h1>
    This header will be red!
  </h1>
</div>
```

If you make the imported style _more specific_, it will have higher precedence over the scoped style:

```css title="make-it-purple.css"
div > h1 {
  color: purple;
}
```
```astro title="MyComponent.astro"
---
import "./make-it-purple.css"
---
<style>
  h1 { color: red }
</style>
<div>
  <h1>
    This header will be purple!
  </h1>
</div>
```

### Import Order

> When importing multiple stylesheets, if any styles have the same specificity, the _last one imported_ wins:

Reviewer: ```suggestion
When importing multiple stylesheets in an Astro component, the CSS rules are evaluated in the order that they are imported. A higher specificity will always determine which styles to show, no matter when the CSS is evaluated. But, when conflicting styles have the same specificity, the _last one imported_ wins:
```
Something like this. I wanted to (a) give this a bit more (even if repeated) context in case anyone jumps to this section right away. And (b) explain what's happening in terms of the order of evaluation (since that's the key factor here). If they _have_ been reading diligently top to bottom, then they KNOW order of evaluation matters when specificity is the same. What they need to know in this section is "In which order does ASTRO evaluate multiple stylesheets?" and I prefer to say that explicitly, matching the context we've set above.

Author: I like this, thanks!

```css title="make-it-purple.css"
div > h1 {
  color: purple;
}
```
```css title="make-it-green.css"
div > h1 {
  color: green;
}
```
```astro title="MyComponent.astro"
---
import "./make-it-green.css"
import "./make-it-purple.css"
---
<style>
  h1 { color: red }
</style>
<div>
  <h1>
    This header will be purple!
  </h1>
</div>
```

If a component imports CSS, that CSS will be imported at the same time, even if the component is never used:
```astro title="PurpleParent.astro"
---
import "./make-it-purple.css"
---
<div>
  <slot/>
</div>
```
```astro title="MyComponent.astro"
---
import "./make-it-green.css"
import PurpleParent from "./PurpleParent.astro";
---
<style>
  h1 { color: red }
</style>
<div>
  <h1>
    This header will be purple!
  </h1>
</div>

> ```

Author: Totally agree, thanks! I originally wanted to use the component as a parent, but then I realized I needed to show that it even brings its CSS when _not_ a parent.

### Link Tags

> During compilation, scoped styles and imported stylesheets are added to the _bottom of the head_ as `<link>` tags, after any other link tags.
> 
> This means that if you load a style sheet via a [link tag](#load-a-static-stylesheet-via-link-tags) in the head, it will always have lower precedence:

Reviewer: And, I do find this a bit confusing too. If this is my `.astro` component as written, there is nothing being added to the head _of my original source component_ that I didn't explicitly write? That's what happens to create the HTML that is sent to the browser, but that doesn't change what's in my actual component file? So, I don't really like that we immediately start talking about turning things into HTML link elements when we really need to tell people when these links as written in their project will be evaluated in comparison to other styling options in their project.

And again, I'd think about rephrasing this in terms of the two factors, specificity and order of appearance, that you've established at the beginning. It's like if we give those rules to follow (Specificity first, then order of appearance), then every explanation only needs speak in those terms. Assuming they know what specificity means, then the real job of this section is to explain Astro's ORDER of evaluating everything. That's what they mostly need to know.

I've rewritten something, which I'm not sure is what you intended, but is what the example seems to want to be explaining.

```suggestion
Style sheets loaded via [link tags](#load-a-static-stylesheet-via-link-tags) in the head are evaluated before any other styles in an Astro component, in the order in which they appear. Therefore, these styles will always have lower precedence than imported styleshees and scoped styles:
```

The other stuff feels more like a "fun fact: ALL styles eventually become stylesheet links in the browser HTML!" if you want to try to include it. But, I don't like it as the description of the *reason* for the styling of this example.

> When Astro bundles and processes your CSS, scoped styles and imported stylesheets are added to the _bottom of the head_ as `<link>` tags, after any other link tags. Your browser then evaluates these link tags in order of their appearance.

Author: Gotcha, I can see how referring to the head of the site introduces more confusion than it explains. I took your suggestion and just swapped the clauses to make it slightly more clear, and removed my unnecessary "always":

> Style sheets loaded via [link tags](#load-a-static-stylesheet-via-link-tags) in the head are evaluated in order, before any other styles in an Astro component. Therefore, these styles will have lower precedence than imported stylesheets and scoped styles:

Because this approach uses the `public/` directory, it skips the normal CSS processing, bundling and optimizations that are provided by Astro. If you need these transformations, use the [Import a Stylesheet](#import-a-local-stylesheet) method above.

## Cascading Order

> Astro components will sometimes have to evaluate multiple CSS "instructions" for the same element. For example, your component might import a CSS stylesheet, include its own `<style>` tag, *and* be rendered inside a layout.

Reviewer: Ooh, I couldn't think of what word I wanted to use instead of "rules" in case that was going to be too repetitive, so I used "instructions" but put in scare quotes to make it a placeholder. Is there a word you like better? Or, if instructions is fine, maybe it should be without the quotation marks?

Author: Or just leave it out:

> `.astro` files will sometimes have to evaluate multiple styles that apply to the same element. For example, your component might import a CSS stylesheet, include its own `<style>` tag, *and* be rendered inside a layout.

Reviewer: Yup, that looks good. Ship it!

Author: I think wherever we use "Astro components", there's a potential for confusion that it doesn't apply to pages, and this is equally true in an intro

Author: Following our convo, I'm going to leave it as "Astro components"!

}
```
```astro title="MyComponent.astro"
---
import "./make-it-purple.css"
---
<style>
  h1 { color: red }
</style>
<div>
  <h1>
    This header will be red!
  </h1>
</div>
```

If you make the imported style _more specific_, it will have higher precedence over the scoped style:

```css title="make-it-purple.css"
div > h1 {
  color: purple;
}
```
```astro title="MyComponent.astro"
---
import "./make-it-purple.css"
---
<style>
  h1 { color: red }
</style>
<div>
  <h1>
    This header will be purple!
  </h1>
</div>
```

### Import Order

When importing multiple stylesheets in an Astro component, the CSS rules are evaluated in the order that they are imported. A higher specificity will always determine which styles to show, no matter when the CSS is evaluated. But, when conflicting styles have the same specificity, the _last one imported_ wins:

```css title="make-it-purple.css"
div > h1 {
  color: purple;
}
```
```css title="make-it-green.css"
div > h1 {
  color: green;
}
```
```astro title="MyComponent.astro"
---
import "./make-it-green.css"
import "./make-it-purple.css"
---
<style>
  h1 { color: red }
</style>
<div>
  <h1>
    This header will be purple!
  </h1>
</div>
```

> If a component imports CSS, that CSS will be imported at the same time, even if the component is never used:

Reviewer: ```suggestion
Importing a component also imports its CSS style rules along with it, even if the component is never used:
```
Realized this didn't actually mention the part about a component importing another component. Could have said, "If an **imported** component imports CSS..." but that starts to get mucky. Tried to rearrange to make it as succinct as possible, to cut down on repeated "component" and "import" as much as possible, but you might have something better.

Reviewer 2: This could get tricky to understand if previously we said that CSS doesn't leak from other components. I was thinking maybe we can include some good practices to go around it. what do you think Sarah and Dan?

Author: We said that _scoped_ styles don't leak - maybe the best practice here is to use Scoped Styles instead?

Author: Oh, and to be sure to import your layout first! Adding

Reviewer: Yes, this "not leaking" tripped me up at first... I almost wrote, WAIT IS THIS TRUE??
 
So to be clear, it's only the IMPORTED styles that carry over with the component, not its `<style>` tag styles?

So, we *could* try to rephrase like "imported styles are transitive" kind of thing... if a style finds its way into your component via importing (whether imported directly via a statment, or imported indirectly via a *component* import), then it's gonna end up in the list of styles that will be evaluated. And, its precedence will depend on the order of the import statement that brought it in: whether you were importing a stylesheet directly, or a component that imported its own stylesheet.

Author: I think "leaking" might be the best way to describe this... tell me what you think of my rephrasing and tip in the latest commit

Reviewer: I like it!

```astro title="PurpleComponent.astro"
---
import "./make-it-purple.css"
---
<div>
  <h1>I import purple CSS.</h1>
</div>
```
```astro title="MyComponent.astro"
---
import "./make-it-green.css"
import PurpleComponent from "./PurpleComponent.astro";
---
<style>
  h1 { color: red }
</style>
<div>
  <h1>
    This header will be purple!
  </h1>
</div>
```

### Link Tags

> Style sheets loaded via [link tags](#load-a-static-stylesheet-via-link-tags), before any other styles in an Astro component. Therefore, these styles will have lower precedence than imported stylesheets and scoped styles:

Reviewer: ```suggestion
Style sheets loaded via [link tags](#load-a-static-stylesheet-via-link-tags) are evaluated before any other styles in an Astro component. Therefore, these styles will have lower precedence than imported stylesheets and scoped styles:
```
